python/src/intracktive/server.py

```diff
@@ -0,0 +1,117 @@
+import logging
+import socket
+import socketserver
+import threading
+from http.server import HTTPServer, SimpleHTTPRequestHandler
+from pathlib import Path
+
+import click
+
+DEFAULT_HOST = "127.0.0.1"
+logging.basicConfig(level=logging.INFO)
+
+
+class ThreadingHTTPServer(socketserver.ThreadingMixIn, HTTPServer):
+    """Handle requests in a separate thread."""
+
+    daemon_threads = True  # Ensure threads close when main thread exits
+
+
+def find_available_port(starting_port=8000):
+    port = starting_port
+    while True:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            if sock.connect_ex((DEFAULT_HOST, port)) != 0:  # Port is free
+                return port
+            port += 1  # Increment to find the next available port
+
+
+def serve_directory(
+    path: Path,
+    host: str = DEFAULT_HOST,
+    port: int = 8000,
+    threaded: bool = True,
+) -> str:
+    """
+    Starts an HTTP server in a background thread to serve a directory, allowing non-blocking execution.
+
+    Parameters
+    ----------
+    path : Path
+        The directory to serve.
+    host : str
+        The host name or IP address, by default 127.0.0.1 (localhost).
+    port : int
+        The port number to serve on, by default 8000.
+    threaded : bool
+        Whether to run the server in a separate thread, by default True.
+
+    Returns
+    -------
+    str
+        The URL of the server.
+    """
+
+    port = find_available_port(port)  # Get an available port
+
+    # Ensure path exists and is a directory
+    if not path.exists() or not path.is_dir():
+        # print(f"The specified path does not exist or is not a directory: {path}")
+        logging.error(
+            "The specified path does not exist or is not a directory: %s", path
+        )
+        return
+
+    class CORSRequestHandler(SimpleHTTPRequestHandler):
```

Comment:
Hi @TeunHuijben, it's unclear why this class needs to be defined inside the function. Could you include it as a comment if that's a requirement?